pkg/operator/ceph/object/objectstore.go

```diff
@@ -758,26 +761,85 @@ func enableRGWDashboard(context *Context) error {
 	if err != nil || errCode != 0 {
 		return errors.Wrapf(err, "failed to create user %q", DashboardUser)
 	}
-	args := []string{"dashboard", "set-rgw-api-access-key", *u.AccessKey}
-	cephCmd := ceph.NewCephCommand(context.Context, context.clusterInfo, args)
-	_, err = cephCmd.Run()
-	if err != nil {
-		return errors.Wrapf(err, "failed to set user %q accesskey", DashboardUser)
-	}
 
-	args = []string{"dashboard", "set-rgw-api-secret-key", *u.SecretKey}
-	cephCmd = ceph.NewCephCommand(context.Context, context.clusterInfo, args)
-	go func() {
-		// Setting the dashboard api secret started hanging in some clusters
-		// starting in ceph v15.2.8. We run it in a goroutine until the fix
-		// is found. We expect the ceph command to timeout so at least the goroutine exits.
-		logger.Info("setting the dashboard api secret key")
-		_, err = cephCmd.RunWithTimeout(applyDashboardKeyTimeout)
+	cephVersion := mgr.CephVersionsNewCephDashboardCommand(context.clusterInfo)
+	// for latest Ceph versions
+	if cephVersion {
+		// Generate a temp file
+		file, err := ioutil.TempFile("", "")
+		if err != nil {
+			return errors.Wrap(err, "failed to generate temp file")
+		}
+		err = ioutil.WriteFile(file.Name(), []byte(*u.AccessKey), 0444)
+		if err != nil {
+			return errors.Wrap(err, "failed to write into file")
+		}
+
+		args := []string{"dashboard", "set-rgw-api-access-key", "-i", file.Name()}
+		cephCmd := ceph.NewCephCommand(context.Context, context.clusterInfo, args)
+		_, err = cephCmd.Run()
+		if err != nil {
+			return errors.Wrapf(err, "failed to set user %q accesskey", DashboardUser)
+		}
+		defer func() {
+			if err := os.Remove(file.Name()); err != nil {
+				logger.Error(err, "failed to clean up file")
+			}
+		}()
+
+		// Generate a temp file
+		file, err = ioutil.TempFile("", "")
+		if err != nil {
+			return errors.Wrap(err, "failed to generate temp file")
+		}
+		err = ioutil.WriteFile(file.Name(), []byte(*u.SecretKey), 0444)
+		if err != nil {
+			return errors.Wrap(err, "failed to write into file")
+		}
+
+		args = []string{"dashboard", "set-rgw-api-secret-key", "-i", file.Name()}
+		cephCmd = ceph.NewCephCommand(context.Context, context.clusterInfo, args)
+		go func() {
+			// Setting the dashboard api secret started hanging in some clusters
+			// starting in ceph v15.2.8. We run it in a goroutine until the fix
+			// is found. We expect the ceph command to timeout so at least the goroutine exits.
+			logger.Info("setting the dashboard api secret key")
+			_, err = cephCmd.RunWithTimeout(applyDashboardKeyTimeout)
+			if err != nil {
+				logger.Errorf("failed to set user %q secretkey. %v", DashboardUser, err)
+			}
+			logger.Info("done setting the dashboard api secret key")
+		}()
+		defer func() {
+			if err := os.Remove(file.Name()); err != nil {
+				logger.Error(err, "failed to clean up file")
+			}
+		}()
```

Comment:
**WARNING**

I think this introduces a race condition. Once the `go func() {...}` starts running in a new goroutine, Go will continue to defer the func to remove the file. Then this function will exit and the deferred `os.Remove(...)` will run, possibly before the `go func()` has completed `cephCmd.RunWithTimeout`.

IMO, I think we should **NOT** run the file-based generation in a goroutine. If this reintroduces the hangs we were seeing earlier, then we will have flaky CI again and can come back here to fix it later.

@travisn / @leseb  thoughts?

Reply:
Alternately, if you make a helper function that (1) creates the temp file, (1a) defers cleanup of the file, (2) runs the CLI command, and (3) cleans up the file in the deferred func, then you can call the helper function like `go helperFunc()`. And that way, the race condition is avoided, and we aren't at risk of CI being flaky again.

Reply:
I think it is better now. @BlaineEXE  PTAL